### ISSUE TITLE
Minor Bug in estimator/tensorflow_estimator/python/estimator/canned/dnn.py

### ISSUE THREAD

**Issue by Anmol-Sharma** [tensorflow/tensorflow#46342]
**System information**
- Model Example : https://www.tensorflow.org/api_docs/python/tf/estimator/DNNClassifier
- OS Platform and Distribution : MacOS 11.1
- TensorFlow installed from : Binary (pip)
- TensorFlow version : 2.4.0
- Python version: 3.8

**Current Behavior : Error**

I was training a DNN Classifier recently and customizing the model as per the example mentioned in the link above.

Here is the code snippet of my model :
```
# Build a DNN with 2 hidden layers with 15 and 10 hidden nodes each.
classifier = tf.estimator.DNNClassifier(
    
    feature_columns=feature_columns, # Specify the list of feature columns
    
    hidden_units=[15, 10], # Specify the list of hidden units
    
    n_classes=2, # Specify the number of classes
    
    # Specify the optimizer
    optimizer = tf.compat.v1.train.ProximalAdagradOptimizer(
      learning_rate=0.02,
      l2_regularization_strength=0.005
    ),
    
    # Specify the dropout
    dropout = 0.2
)
```

However, while training the defined model, I get the following error : 
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-dcc1423bd4e7> in <module>
      1 # Train the Model.
----> 2 classifier.train( input_fn=train_input_fn )

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/estimator.py in train(self, input_fn, hooks, steps, max_steps, saving_listeners)
    347 
    348       saving_listeners = _check_listeners_type(saving_listeners)
--> 349       loss = self._train_model(input_fn, hooks, saving_listeners)
    350       logging.info('Loss for final step: %s.', loss)
    351       return self

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/estimator.py in _train_model(self, input_fn, hooks, saving_listeners)
   1173       return self._train_model_distributed(input_fn, hooks, saving_listeners)
   1174     else:
-> 1175       return self._train_model_default(input_fn, hooks, saving_listeners)
   1176 
   1177   def _train_model_default(self, input_fn, hooks, saving_listeners):

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/estimator.py in _train_model_default(self, input_fn, hooks, saving_listeners)
   1201           self._get_features_and_labels_from_input_fn(input_fn, ModeKeys.TRAIN))
   1202       worker_hooks.extend(input_hooks)
-> 1203       estimator_spec = self._call_model_fn(features, labels, ModeKeys.TRAIN,
   1204                                            self.config)
   1205       global_step_tensor = tf.compat.v1.train.get_global_step(g)

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/estimator.py in _call_model_fn(self, features, labels, mode, config)
   1161 
   1162     logging.info('Calling model_fn.')
-> 1163     model_fn_results = self._model_fn(features=features, **kwargs)
   1164     logging.info('Done calling model_fn.')
   1165 

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/canned/dnn.py in _model_fn(features, labels, mode, config)
    748     def _model_fn(features, labels, mode, config):
    749       """Call the defined shared dnn_model_fn_v2."""
--> 750       return dnn_model_fn_v2(
    751           features=features,
    752           labels=labels,

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/canned/dnn.py in dnn_model_fn_v2(***failed resolving arguments***)
    572   # relies on global step as step counter.
    573   if mode == ModeKeys.TRAIN:
--> 574     optimizer = optimizers.get_optimizer_instance_v2(optimizer)
    575     optimizer.iterations = tf.compat.v1.train.get_or_create_global_step()
    576 

/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/canned/optimizers.py in get_optimizer_instance_v2(opt, learning_rate)
    141     opt = opt()
    142   if not isinstance(opt, optimizer_v2.OptimizerV2):
--> 143     raise ValueError(
    144         'The given object is not a tf.keras.optimizers.Optimizer instance.'
    145         ' Given: {}'.format(opt))

ValueError: The given object is not a tf.keras.optimizers.Optimizer instance. Given: <tensorflow.python.training.proximal_adagrad.ProximalAdagradOptimizer object at 0x13b9bea90>
```

**Expected behavior**

The model is demanding the optimizer to be an instance of the Optimizer of the Keras API. This should not be the case, this is perfectly fine code.

So I did some digging inside the code in **dnn.py** and found a minor issue here :

https://github.com/tensorflow/estimator/blob/49efa0187fcfefa1660f0aae745a4c3c0793f752/tensorflow_estimator/python/estimator/canned/dnn.py#L574

Code There (Line 574) :
```
if mode == ModeKeys.TRAIN:
    optimizer = optimizers.get_optimizer_instance_v2(optimizer)
    optimizer.iterations = tf.compat.v1.train.get_or_create_global_step()
```

Here we can clearly see, that we are calling the `get_optimizer_instance_v2` without checking the type of the optimizer. If our optimizer is of type **tf.compat.v1.train.Optimizer**, this call will fail, which is precisely what is happening.

**Method to reproduce**

Create a DNN Model with the `tf.estimator` api and specify optimizer of the type : `tf.compat.v1.train.ProximalAdagradOptimizer`

Please verify this issue on your end.

If verified, I will generate a pull request fixing this issue.

**Comment by amahendrakar** [comment]
@Anmol-Sharma,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and the dataset you are using. 

> If verified, I will generate a pull request fixing this issue.

Also, please feel free to submit a PR regarding this. Thanks!

**Comment by Anmol-Sharma** [comment]
@amahendrakar You can use the code below as an example to trouble-shoot. It is similar to what I'm doing. I've verified the issue with this script as well.

Just use the Titanic dataset from kaggle and try to run this on this script.

```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# For Tensorflow
import tensorflow as tf
from tensorflow.estimator import LinearRegressor, DNNClassifier

# For Splitting Training / Testing Data
from sklearn.model_selection import train_test_split
import time


# Titanic Dataset from Kaggle
df = pd.read_csv("../../datasets/Titanic.csv")


# Drop the non-useful columns :- PassengerId, Name, Ticket Nr
df.drop(columns=["PassengerId", "Name", "Ticket"], inplace = True)


# Check of null-values
df.isna().sum()

# Cabin has a lot of null values, so we discard that column
df.drop(columns = ["Cabin"], inplace = True)

# Group Age by Sex and Age and Set the mean age by sex for the missing values.
res = df.groupby("Sex")["Age"].mean()
idx = res.index
mapping = dict(zip(idx, res))

# Add data for missing Age
df.loc[df["Age"].isnull() == True, "Age"] = df.loc[df["Age"].isnull() == True, "Sex"].map(lambda x : mapping[x])

# Add data for missing Embarked : Set the Embarked to the most common value
df.loc[df["Embarked"].isnull() == True, "Embarked"] = df["Embarked"].value_counts()[0]

df.drop(df[df.Embarked == 644].index.values, inplace = True)


output = df.Survived
df.drop(columns = ["Survived"], inplace = True)

train_x, test_x, train_y, test_y = train_test_split(df, output, test_size=0.1, random_state=int(time.time()))


CATEGORICAL_COLUMNS = ["Pclass", "Sex", "Embarked"]
NUMERIC_COLUMNS = ['Age', 'Fare']

# Create the list of feature columns
feature_columns = []
for feature_name in CATEGORICAL_COLUMNS:
    vocabulary = train_x[feature_name].unique()
    # For Categorical column we are specify categorical column with vocab list.
    # There are other types for categorical data as well. Refer to TF documentation.
    col = tf.feature_column.categorical_column_with_vocabulary_list(feature_name, vocabulary)
    feature_columns.append(tf.feature_column.indicator_column(col))


# For Non categorical data, we define the column type to be numeric_column
for feature_name in NUMERIC_COLUMNS:
    feature_columns.append(tf.feature_column.numeric_column(feature_name, dtype=tf.float32))

# Define an inut function, which will generate the dataset for our model
def make_input_fn(data_df, label_df, num_epochs=10, shuffle=True, batch_size=32):
    
    # A helper function to be returned
    def input_function():
        ds = tf.data.Dataset.from_tensor_slices((dict(data_df), label_df))
        if shuffle:
            ds = ds.shuffle(1000)
        ds = ds.batch(batch_size).repeat(num_epochs)
        return ds
    
    return input_function

train_input_fn = make_input_fn(train_x, train_y, batch_size = 2)
eval_input_fn = make_input_fn(test_x, test_y, num_epochs=1, shuffle=False)

# Build a DNN with 2 hidden layers with 15 and 10 hidden nodes each.
classifier = tf.estimator.DNNClassifier(
    
    feature_columns=feature_columns, # Specify the list of feature columns
    
    hidden_units=[15, 10], # Specify the list of hidden units
    
    n_classes=2, # Specify the number of classes
    
    # Specify the dropout
    dropout = 0.2
)

# Train the Model.
classifier.train( input_fn=train_input_fn )

eval_result = classifier.evaluate(input_fn=eval_input_fn)

print('\nTest set accuracy: {accuracy:0.3f}\n'.format(**eval_result))
```

Also, I think the PR will have to be issued for the same in the [tensorflow/estimator](https://github.com/tensorflow/estimator) repository, as the core code in dnn.py is present there.
I will submit a PR there for this fix and link it here to this issue.

**Comment by amahendrakar** [comment]
@Anmol-Sharma,
I did not face any errors while running the code with TF v2.4. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/c9e179b16e30b3ca6d7d1696c3f489f6/46342.ipynb). 

Could you please try running the code in a new virtual environment and check if you are facing the same error. Thanks!

**Comment by Anmol-Sharma** [comment]
@amahendrakar 

That is because in your code, one extra line of optimizer specification is missing. This 👇 
```
# Specify the optimizer
    optimizer = tf.compat.v1.train.ProximalAdagradOptimizer(
      learning_rate=0.02,
      l2_regularization_strength=0.005
    ),
```

Add this to your model declaration and then try to train, you will get the error.

**Comment by amahendrakar** [comment]
@Anmol-Sharma,
Thank you for the update. 

@jvishnuvardhan,
I was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/4fe23c2686e381e2858d607749047d23/46342.ipynb). Thanks!

**Comment by Anmol-Sharma** [comment]
@amahendrakar 👍 

I've issued a PR, fixing the issue in the tensorflow/estimator repo. Have a look when you can.

**Comment by zhjunqin** [comment]
I found similar issue when use `tf.compat.v1.train.AdagradOptimizer` and multi_head in estimator in TF estimator 2.4

The PR https://github.com/tensorflow/estimator/pull/59 not solve my issue.


```
  File "/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/head/multi_head.py", line 491, in create_estimator_spec
    train_op = base_head.create_estimator_spec_train_op(
  File "/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/head/base_head.py", line 904, in create_estimator_spec_train_op
    validate_v2_optimizer(optimizer)
  File "/usr/local/lib/python3.8/site-packages/tensorflow_estimator/python/estimator/head/base_head.py", line 667, in validate_v2_optimizer
    raise ValueError(
ValueError: The given optimizer is not a tf.keras.optimizers.Optimizer instance. Given: <tensorflow.python.training.adagrad.AdagradOptimizer object at 0x7f4f8d11d070>
```

So I follow code in https://github.com/tensorflow/estimator/blob/r1.14/tensorflow_estimator/python/estimator/head/multi_head.py#L427

Change code in function `create_estimator_spec_train_op` like below:

```
          if isinstance(optimizer, optimizer_v2.OptimizerV2):
            validate_v2_optimizer(optimizer)
            validate_trainable_variables(trainable_variables)
            # optimizer_v2.get_updates always returns a list, and the first
            # element is the train_op.
            train_op = optimizer.get_updates(
                loss, trainable_variables)[0]
          else:
            train_op = optimizer.minimize(
                regularized_training_loss,
                global_step=training_util.get_global_step())
```

**Comment by sushreebarsa** [comment]
@Anmol-Sharma Is this still an issue ?
Estimators are not recommended for new code. Estimators  can behave unexpectedly, especially when combined with TF 2 code. Could you please refer to this [link](https://www.tensorflow.org/api_docs/python/tf/estimator/DNNClassifier) for more details  and let us know if it helps?
Thanks!

**Comment by Anmol-Sharma** [comment]
Hi @sushreebarsa I don't this is still an issue. At that time when I opened this issue, it was present and I raised a fix as well [Here](https://github.com/tensorflow/estimator/pull/59) but due to very large merge conflicts (when the time to merge came) the changes were skipped.

So, I think this ticket can be closed, as mixing of Estimators with TF2 may not be a very good idea, so I'm closing this.
